src/debug.c

```diff
@@ -14,20 +23,104 @@ void gssntlm_debug_init(void)
 {
     char *env;
 
+    pthread_mutex_lock(&debug_mutex);
+
     env = secure_getenv("GSSNTLMSSP_DEBUG");
     if (env) {
         debug_fd = fopen(env, "a");
         if (debug_fd) gssntlm_debug_enabled = true;
     }
     gssntlm_debug_initialized = true;
+
+    pthread_mutex_unlock(&debug_mutex);
 }
 
 void gssntlm_debug_printf(const char *fmt, ...)
 {
     va_list ap;
 
-    va_start(ap, fmt);
-    vfprintf(debug_fd, fmt, ap);
-    va_end(ap);
-    fflush(debug_fd);
+    pthread_mutex_lock(&debug_mutex);
+
+    if (debug_fd) {
+        va_start(ap, fmt);
+        vfprintf(debug_fd, fmt, ap);
+        va_end(ap);
+        fflush(debug_fd);
+    }
+
+    pthread_mutex_unlock(&debug_mutex);
+}
+
+static int gssntlm_debug_enable(const char *filename)
+{
+    FILE *old_debug_fd = debug_fd;
+    FILE *new_debug_fd = NULL;
+    int ret = 0;
+
+    pthread_mutex_lock(&debug_mutex);
+
+    new_debug_fd = fopen(filename, "a");
+    if (new_debug_fd) {
+        gssntlm_debug_enabled = true;
+    } else {
+        gssntlm_debug_enabled = false;
+        ret = errno;
+    }
+
+    debug_fd = new_debug_fd;
+
+    if (old_debug_fd != NULL) {
+        fclose(old_debug_fd);
+    }
+
+    pthread_mutex_unlock(&debug_mutex);
+
+    return ret;
+}
+
+static int gssntlm_debug_disable(void)
+{
+    FILE *old_debug_fd = debug_fd;
+    int ret = 0;
+
+    pthread_mutex_lock(&debug_mutex);
+
+    gssntlm_debug_enabled = false;
+    debug_fd = NULL;
+
+    if (old_debug_fd != NULL) {
+        ret = fclose(old_debug_fd);
+    }
+
+    pthread_mutex_unlock(&debug_mutex);
+
+    return ret;
+}
+
+gss_OID_desc gssntlm_debug_oid = {
+    GSS_NTLMSSP_DEBUG_OID_LENGTH,
+    discard_const(GSS_NTLMSSP_DEBUG_OID_STRING)
+};
+
+int gssntlm_debug_invoke(gss_buffer_t value)
+{
+    char filename[PATH_MAX] = { 0 };
+
+    gssntlm_debug_initialized = true;
+
+    if (value->length > PATH_MAX - 1) {
+        return EINVAL;
```

Comment:
In case if error is returned here, init was not done but gssntlm_debug_initialized is still set to true (few lines above).
Should we move setting gssntlm_debug_initialized below this error handling?

Reply:
probably